evaluate.go

```diff
@@ -237,34 +237,83 @@ func evaluateNotPresent(ptr pointerstructure.Pointer, datum interface{}) bool {
 	return reflect.ValueOf(val).Kind() == reflect.Map
 }
 
-func evaluateMatchExpression(expression *grammar.MatchExpression, datum interface{}, opt ...Option) (bool, error) {
+func getValue(datum interface{}, path []string, opt ...Option) (interface{}, bool, error) {
 	opts := getOpts(opt...)
-	ptr := pointerstructure.Pointer{
-		Parts: expression.Selector.Path,
-		Config: pointerstructure.Config{
-			TagName:                 opts.withTagName,
-			ValueTransformationHook: opts.withHookFn,
-		},
-	}
-	val, err := ptr.Get(datum)
-	if err != nil {
-		if errors.Is(err, pointerstructure.ErrNotFound) {
-			// Prefer the withUnknown option if set, otherwise defer to NotPresent
-			// disposition
-			switch {
-			case opts.withUnknown != nil:
-				err = nil
-				val = *opts.withUnknown
-			case evaluateNotPresent(ptr, datum):
-				return expression.Operator.NotPresentDisposition(), nil
+	var val interface{}
+	var found bool
+	if len(path) != 0 {
+		for i := len(opts.withLocalVariables) - 1; i >= 0; i-- {
+			name := path[0]
+			lv := opts.withLocalVariables[i]
+			if name == lv.name {
+				if len(lv.path) == 0 {
+					// This local variable is a key or an index and we know its
+					// value without having to call pointerstructure, we stop
+					// here.
+					val = lv.value
+					if len(path) > 1 {
+						first := pointerstructure.Pointer{Parts: []string{name}}
+						full := pointerstructure.Pointer{Parts: path}
+						return nil, false, fmt.Errorf("%s references a %T so %s is invalid", first.String(), val, full.String())
+					}
+					found = true
+					break
```

Comment:
Couldn't this just be an early return and the subsequent `if !found` conditional removed to always assume it wasn't found?

Reply:
Yes, at that time I wanted to support composite map keys but since this is not supported for the time being we can exit early.